pkg/protocol/engine/congestioncontrol/scheduler/drr/issuerqueue.go

```diff
@@ -123,6 +123,10 @@ func (q *IssuerQueue) Front() *blocks.Block {
 
 // PopFront removes the first ready block from the queue.
 func (q *IssuerQueue) PopFront() *blocks.Block {
+	if q.inbox.Len() == 0 {
```

Comment:
Isn't that check a bit racy with the `heap.Pop` ?

Reply:
Access to the buffer (`IssuerQueue`) is protected by `s.bufferMutex` in the scheduler - that guarantees that only one goroutine can call scheduler methods that work with the buffer at a time. The same applies to your other comments.